core/embed/extmod/modtrezorutils/modtrezorutils.c

```diff
@@ -117,6 +119,49 @@ STATIC mp_obj_t mod_trezorutils_halt(size_t n_args, const mp_obj_t *args) {
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_trezorutils_halt_obj, 0, 1,
                                            mod_trezorutils_halt);
 
+/// def firmware_hash(challenge: bytes | None) -> bytes:
+///     """
+///     Computes the Blake2s hash of the firmware with an optional challenge as
+///     the key.
+///     """
+STATIC mp_obj_t mod_trezorutils_firmware_hash(size_t n_args,
+                                              const mp_obj_t *args) {
+  BLAKE2S_CTX ctx;
+  mp_buffer_info_t chal = {0};
+  if (n_args > 0 && args[0] != mp_const_none) {
+    mp_get_buffer_raise(args[0], &chal, MP_BUFFER_READ);
+  }
+
+  if (chal.len != 0) {
+    if (blake2s_InitKey(&ctx, BLAKE2S_DIGEST_LENGTH, chal.buf, chal.len) != 0) {
+      mp_raise_msg(&mp_type_ValueError, "Invalid challenge.");
+    }
+  } else {
+    blake2s_Init(&ctx, BLAKE2S_DIGEST_LENGTH);
+  }
+
+  for (int i = 0; i < FIRMWARE_SECTORS_COUNT; i++) {
+    uint8_t sector = FIRMWARE_SECTORS[i];
+    uint32_t size = flash_sector_size(sector);
+    const void *data = flash_get_address(sector, 0, size);
+    if (data == NULL) {
+      mp_raise_msg(&mp_type_RuntimeError, "Failed to read firmware.");
+    }
+    blake2s_Update(&ctx, data, size);
+  }
+
+  vstr_t vstr = {0};
+  vstr_init_len(&vstr, BLAKE2S_DIGEST_LENGTH);
+  if (blake2s_Final(&ctx, vstr.buf, vstr.len) != 0) {
+    vstr_clear(&vstr);
+    mp_raise_msg(&mp_type_RuntimeError, "Failed to finalize firmware hash.");
+  }
+
+  return mp_obj_new_str_from_vstr(&mp_type_bytes, &vstr);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_trezorutils_firmware_hash_obj, 0,
```

Comment:
this should actually be a one-argument (not variadic) function, checking for None

Reply:
But then calling `utils.firmware_hash()` will result in `TypeError: function takes 1 positional arguments but 0 were given`. I meant the definition to be `/// def firmware_hash(challenge: bytes | None = None) -> bytes:`.

Reply:
ISTM nobody is calling it as `utils.firmware_hash()` anyway so it makes things a bit simpler to just require the argument.